### PR TITLE
Authenticating microservices requests in Kubernetes

### PR DESCRIPTION
**Pull request by amitsaha**
This PR creates the content for the article "Authenticating microservices requests in Kubernetes".

### CODE REVIEW

File: src/microsAuthentication/content.md [hunk 1]
```diff
@@ -0,0 +1,1118 @@
...
+
+1. An API
+1. A data store
+1. A secret store
+
+You might want the secret store only to reply to requests to the API and reject any request from the data store.
+
+How would the secret store decide to authenticate or deny the request?
+
+A popular approach is to request and pass identity tokens to every call within services. So instead of
```
Comment (danielepolencic) [inline src/microsAuthentication/content.md:12]: I know this sounds weird, but I have a single sentence per paragraph.

The content is spaced and easier to read.

File: src/microsAuthentication/content.md [hunk 2]
```diff
@@ -0,0 +1,1118 @@
...
+1. A secret store
+
+You might want the secret store only to reply to requests to the API and reject any request from the data store.
+
+How would the secret store decide to authenticate or deny the request?
+
+A popular approach is to request and pass identity tokens to every call within services. So instead of 
+issuing a request from the API service or the data store to the secret store directly, you might need 
+to go through an Authorisation service first, authenticate and retrieve a token that is passed along 
+to your request. That way, when the data store gets a token and presents it to the secret store,
```
Comment (danielepolencic) [inline src/microsAuthentication/content.md:15]: when the data store ~gets a token and~ presents the token to

File: src/microsAuthentication/content.md [hunk 3]
```diff
@@ -0,0 +1,1118 @@
...
+
+You might want the secret store only to reply to requests to the API and reject any request from the data store.
+
+How would the secret store decide to authenticate or deny the request?
+
+A popular approach is to request and pass identity tokens to every call within services. So instead of 
+issuing a request from the API service or the data store to the secret store directly, you might need 
+to go through an Authorisation service first, authenticate and retrieve a token that is passed along 
+to your request. That way, when the data store gets a token and presents it to the secret store, 
+it is rejected, but if the API service presents one, it will be accepted.
```
Comment (danielepolencic) [inline src/microsAuthentication/content.md:16]: but if the API service presents the same token, ...

File: src/microsAuthentication/content.md [hunk 4]
```diff
@@ -0,0 +1,1118 @@
...
+
+![Three apps](image1.png)
+
+You have several options when it comes to Authorisation server:
+
+- You could use static tokens that don’t expire (and perhaps hardcoding them in the applications that
```
Comment (danielepolencic) [inline src/microsAuthentication/content.md:22]: (and perhaps hardcoding them in the applications or loading them at startup from an environment variable).

The fact that the auth server is not there is implicit, I guess?
I'm trying to break down the long sentence in the parenthesis.

File: src/microsAuthentication/content.md [hunk 5]
```diff
@@ -0,0 +1,1118 @@
...
+to your request. That way, when the data store gets a token and presents it to the secret store, 
+it is rejected, but if the API service presents one, it will be accepted.
+
+![Three apps](image1.png)
+
+You have several options when it comes to Authorisation server:
+
+- You could use static tokens that don’t expire (and perhaps hardcoding them in the applications that 
+  communicate with each other and removing the need for an authorisation server at all)
+- You could use oAuth by setting up an internal oAuth server
```
Comment (danielepolencic) [inline src/microsAuthentication/content.md:24]: All bullet points should terminate with a full stop.

File: src/microsAuthentication/content.md [hunk 6]
```diff
@@ -0,0 +1,1118 @@
...
+- You could roll out your own authorisation mechanism such as mutual TLS certificates
+
+All authorisation servers have to do is to:
+
+1. Authenticate the caller -  The caller should have an identity that it’s part of the network and  verifiable
```
Comment (danielepolencic) [inline src/microsAuthentication/content.md:29]: change curly apostrophe with straight apostrophe

File: src/microsAuthentication/content.md [hunk 7]
```diff
@@ -0,0 +1,1118 @@
...
+1. Generate a token with a limited scope, validity and the desired audience
+1. Validate a token - The called service would call the Authorisation server with the provided token and accept the request if the validation is successful and reject it otherwise
+
+Examples of Authorisation servers are tools such as [Keycloak]() or [Dex]().
```
Comment (danielepolencic) [inline src/microsAuthentication/content.md:33]: Missing links

File: src/microsAuthentication/content.md [hunk 8]
```diff
@@ -0,0 +1,1118 @@
...
+
+You might have not noticed, but Kubernetes offers the same primitives of the authorisation service within the 
+cluster with Service Account, Roles and RoleBindings.
+
+You assign identities using Service Accounts. Service accounts are linked to roles to define what
```
Comment (danielepolencic) [inline src/microsAuthentication/content.md:38]: roles that grant access to resources

File: src/microsAuthentication/content.md [hunk 9]
```diff
@@ -0,0 +1,1118 @@
...
+resources they have access to. 
+
+Tokens are generated as soon as you create the Service Account and stored in a Secret
```
Comment (danielepolencic) [inline src/microsAuthentication/content.md:41]: missing full stop.

File: src/microsAuthentication/content.md [hunk 10]
```diff
@@ -0,0 +1,1118 @@
...
+
+You use the secret as a mechanism to authenticate to the API and make requests. You only receive 
+successful replies for the resources you are entitled to consume
+
+Could you use Service Accounts as a mechanism to authenticate requests between apps in the cluster?
```
Comment (danielepolencic) [inline src/microsAuthentication/content.md:46]: Before this question, we could insert:

> Kubernetes offers the same primitives to authentication requests just like Keycloak and Dex.
> _So what if the Kubernetes API could be used as an Authorisation server?_
> Could you use Service Accounts as a mechanism to authenticate requests between apps in the cluster?
> Let's try that

File: src/microsAuthentication/content.md [hunk 11]
```diff
@@ -0,0 +1,1118 @@
...
+
+1. We will refer to these services as the API service (API) and the Secret store 
+1. They are written in the Go programming language and they communicate via HTTP 
+1. Each service will run in its own namespaces and use dedicated service accounts
+1. The secret store requires that any client communicating to it identifies itself 
+   via a valid service account token passed as a HTTP header
+
+## Setting up
+
+You will need access to a Kubernetes cluster with the ServiceAccountVolume projection feature enabled.
```
Comment (danielepolencic) [inline src/microsAuthentication/content.md:64]: ServiceAccountVolume projection -> link to the official docs.

File: src/microsAuthentication/content.md [hunk 12]
```diff
@@ -0,0 +1,1118 @@
...
+   via a valid service account token passed as a HTTP header
+
+## Setting up
+
+You will need access to a Kubernetes cluster with the ServiceAccountVolume projection feature enabled. 
+You will learn more about this feature later on in the article. The ServiceAccountVolume projection 
+requires the Kubernetes API server to be started with certain specific API flags. The support for 
+different Kubernetes providers may vary.
+
+However, you can instead use the latest version of [minikube](). From your terminal, run:
```
Comment (danielepolencic) [inline src/microsAuthentication/content.md:69]: missing link.

**Review by danielepolencic — APPROVED** [review]
I feel like some of these comments are tiny, I can just amend them myself. I'm creating more work for you here.
I will raise a PR against your repo with some of the changes.
I think we could have saved the back and forth with a local branch.
Maybe next time.

Approving, but raising a PR on yours.

**Comment by amitsaha** [comment]
> I feel like some of these comments are tiny, I can just amend them myself. I'm creating more work for you here.
> I will raise a PR against your repo with some of the changes.
> I think we could have saved the back and forth with a local branch.
> Maybe next time.
> 
> Approving, but raising a PR on yours.

Thanks, did you raise a PR already? I cannot see any.

**Comment by danielepolencic** [comment]
I did not. I think it's easier if you push a local branch and we work together on that

**Comment by amitsaha** [comment]
Sure. Do you want to merge this PR? Then I can create a branch and push
where we can collaborate? Or did you mean something else?

**Comment by danielepolencic** [comment]
We can merge this PR, but then I'm not sure how I will able to comment on it.
I think you could push the branch as a local branch.
If you can, add the amendments (which otherwise I will have to re-do).

**Comment by amitsaha** [comment]
> On 11 Nov 2020, at 12:33 pm, Daniele Polencic <notifications@github.com> wrote:
> 
> 
> We can merge this PR, but then I'm not sure how I will able to comment on it.
> I think you could push the branch as a local branch.
> If you can, add the amendments (which otherwise I will have to re-do).
> 
> 
How about I do this?

1. I push your suggested amendments to my fork’s branch (same as the PR)
2. (1) will update the PR automatically

Does that sound close to what you had in mind?

**Comment by danielepolencic** [comment]
that's ok.

But once you have the changes, can you push the same branch to this repo as well?
We close the current PR and raise a local one.

File: src/microsAuthentication/content.md (hunk 1 above)
Comment (amitsaha) [reply src/microsAuthentication/content.md:12]: What if we have a linter rule to detect this?

**Comment by amitsaha** [comment]
Closing this in favor of https://github.com/learnk8s/learnk8s.io/pull/327

**Pull request closed without merging**